README.md
[![Stories in Ready](https://badge.waffle.io/sebobo/jquery.smallipop.png?label=ready&title=Ready)](https://waffle.io/sebobo/jquery.smallipop)
smallipop
=============

[Small Improvements](http://www.small-improvements.com) info popup

This jQuery plugin has been created because all the tooltips we tested didn't meet our requirements.

We wanted:

 * **Pure css** for the design. Fallbacks for older browsers and no images.
 * **CSS Animations** are supported.
 * **Custom themes** for different use cases.
 * The popup trigger **can be positioned anywhere**. The popup will be displayed at the body root, so the position won't be affected by your layout.
 * **Automatic orientation** whereever the popup appears. The popup will try to stay in the visible area of the screen, even when scrolling or at the edges.
 * **Only a single popup** element in the document. Some plugins create a hidden popup for each trigger.
 * **Custom options** for each trigger. You can have 20 different popups with 20 different themes on one page if you like.
 * **Small**. Only ~ 3.5KB minified and ~ 2KB gziped.

We are starting to use this plugin for a lot of things, so watch for new releases at [github](https://github.com/Sebobo/jquery.smallipop).

See the [documentation page](http://sebobo.github.com/jquery.smallipop/) for a live demo and examples.

Do you like this project?
[Buy me a beer](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=TSN2TDYNKZHF4)

Installation
------------

### Prequisites

 * [jQuery](http://www.jquery.com) - 1.5.2 or better
 * [Modernizr](http://www.modernizr.com) - This library tests the browser for feature support and adds classes to the body tag. We use this for css fallbacks in our themes.

Both are also provided in the `lib` folder.

If you don't want to use the `Modernizr` library you can remove the `.cssgradients`, `.borderradius`, `.rgba` and `.boxshadow` classes in `css/jquery-smallipop.css`.

### Required files

Copy `lib/jquery-smallipop.js` to your javascript folder.
Copy `css/jquery-smallipop.css` to your css folder.


Usage
-----

If you like demos more than a boring documentation see the `index.html` file and play with it.

The plugin can be called with jQuery in different ways.

### Standard call with default theme and settings:

    $('.myElement').smallipop();

#### Popup content and markup

If `myElement` is a `<a>` tag, the elements `title` attribute is used as popup text:

    <a class="myElement" href="#" title="Some hint text">
        Sample link
    </a>

You can also provide the text in javascript for all selected elements:

    $('.myElement').smallipop({}, 'This is my special hint');

If `myElement` contains an element with the `smallipop-hint` class, it's content is copied into the popup when displayed.
This can be any markup content you like. Be careful when using floating elements as content,
they need a clearfix afterwards or jQuery is unable to get the correct size for the popup.

    <div class="myElement">
        Sample link
        <span class="smallipop-hint">
           <h6>My bubble title</h6>
           <p>Some text for the bubble</p>
        </span>
    </div>

The plugin first checks if the text has been provided in the javascript call.
If not then the `title` attribute will be used, if the trigger is a link.
If not the element will be checked if it contains another element with the `smallipop-hint` class, and it's html content will be used.

#### Using a different theme:

    $('.myElement').smallipop({
        theme: 'white'
    });

#### Hide the popup trigger while displaying the popup:

    $('.myElement').smallipop({
        hideTrigger: true
    });

#### Provide the hint in javascript:

    $('.myElement').smallipop({}, 'This is my special hint');

#### Changing the default theme for all subsequently created popups:

    $.smallipop.defaults.theme = 'black'

You can change the default value for all other options as well in this way.


### Options

 * popupOffset: horizontal offset for the popup. Default is `31`.
 * popupYOffset: vertical offset for the popup. Default is `0`.
 * popupDistance: vertical distance when the popup appears and disappears. Default is `20`.
 * hideTrigger: hide the trigger when the popup is shown. Default is `false`.
 * theme: `black`, `orange`, `blue`, `white` and `default` are included in the css file. Default is `default`.
 * infoClass: Class in an element which contains markup content for the popup. Default is `smallipop-hint`.
 * popupDelay: How much time in milliseconds you have to hover on an element before the popups shows up. Default is `100`.
 * triggerAnimationSpeed: How fast the trigger fades in and out when `hideTrigger` is enabled, default is `150`.
 * popupAnimationSpeed: How much time the popup needs to reach it's final animation position and opacity in milliseconds. Default is `200`.
 * invertAnimation: The popup will move up when an element is hovered and further up when fading out. If you set this to true, the popup will move down when fading out. Default is `false`.
 * horizontal: The popup will be positioned left or right of the trigger. Default is `false`.

Editing
-------

Read this chapter if you want to modify or extend smallipop.

The plugin is written in [coffeescript](http://jashkenas.github.com/coffee-script/) and the css with [sass](http://sass-lang.com/).
The sources are provided in the `src` and `scss` folders.

So you can either work with the compiled `.js` and `.css` files in your project or use the coffeescript and sass files.

I have provided a combined watcher script `watcher.py` which starts the two watcher daemons, when your editing the files.
This requires the installation of python, coffeescript and sass. You can find very good installation instructions on the project homepages.

The generated css for the themes is quite long. Remove any themes you don't need.

The css used for the popup arrow is a bit tricky. If you want to change it, you need to know how css borders are rendered or you use images instead.


Feedback
--------

Please send me an [email](sebastian@helzle.net) or a tweet @sebobo with any feedback you have.

This plugin was my first attempt at a custom tooltip, coffeescript and scss, so any ideas for improvement are welcome.


Contributing
------------

Contribute!


CHANGELOG.md
# Changelog

## 0.6.4
* [TASK] Add composer support - thanks to Daniel Herrmann

## 0.6.3
* [TASK] Don't create smallipop instances when selector is empty.
* [TASK] Divided scss into separate files so they can be included more easily.
* [TASK] Added package.json with correct dependencies for developing.

## 0.6.2
* [BUGFIX] Uppercase nodeName before checking if element is HTML - thanks to Christian Weiske
* [TASK] Added waffle.io badge
* [TASK] Added bower.json
* [BUGFIX] Hint class was wrong in readme
* [BUGFIX] Tests are running again

## 0.6.1
* Wrong position of arrow when scrolling and smallipop realigns itself
* Mouseout for trigger works again

## 0.6.0
* Done a lot of refactoring. Minified versions of js and css lost some KB.
* Tooltip arrows are now using :after and :before pseudo elements. Thanks to https://github.com/zalog for this patch. These arrows won't show up when using IE7 and earlier.
* All css classes are now using dashes. This simplifies a lot of things and makes the plugin more consistent. You should check your custom themes.
* The default inline hint class is now smallipop-hint instead of smallipopHint. But this is still available as option, so you can use your old smallipopHint class.


tests/tests.js

/*!
Test suite for jQuery Smallipop
 */
var defaultDelay, delayCall;

$('.smallipop').smallipop({
  theme: 'black',
  touchSupport: false
});

$('.run-tour').click(function() {
  return $('.smallipop-tour').smallipop('tour');
});

delayCall = function(delay, callback) {
  return setTimeout(callback, delay);
};

defaultDelay = 1000;


/*
Test the core plugin features
 */

module('core');

test('Smallipop exists', function() {
  var smallipop;
  expect(2);
  smallipop = $('.smallipop-instance:first');
  equal(smallipop.length, 1, 'Smallipop instance should exist');
  return equal(smallipop.attr('id'), 'smallipop1', 'First smallipop should have id 1');
});


/*
Test interaction with the plugins ui elements
 */

module('interaction');

asyncTest('Show and hide tooltip by interaction', function() {
  var smallipop, trigger;
  expect(2);
  smallipop = $('#smallipop1');
  trigger = $('.smallipop:first');
  trigger.mouseover();
  return delayCall(defaultDelay, function() {
    ok(smallipop.is(':visible'), 'Smallipop should be visible');
    $(document).click();
    return delayCall(defaultDelay, function() {
      equal(smallipop.css('display'), 'none', 'Smallipop should now be hidden');
      return start();
    });
  });
});


/*
Test direct calls to the plugins api
 */

module('api');

asyncTest('Show and hide tooltip with api', function() {
  var smallipop, trigger;
  expect(2);
  smallipop = $('#smallipop1');
  trigger = $('.smallipop:first');
  trigger.smallipop('show');
  return delayCall(defaultDelay, function() {
    ok(smallipop.is(':visible'), 'Smallipop should be visible');
    trigger.smallipop('hide');
    return delayCall(defaultDelay, function() {
      equal(smallipop.css('display'), 'none', 'Smallipop should now be hidden');
      return start();
    });
  });
});

asyncTest('Change tooltip content with api', function() {
  var newHint, oldHint, smallipop, trigger;
  expect(2);
  smallipop = $('#smallipop1');
  trigger = $('.smallipop:first');
  oldHint = trigger.data('smallipop').hint;
  newHint = 'some fancy new hint';
  notEqual(oldHint, newHint, 'Old and new content are different');
  trigger.smallipop('show');
  trigger.smallipop('update', newHint);
  return delayCall(defaultDelay, function() {
    equal($('.smallipop-content', smallipop).text(), newHint, 'Hint should have changed');
    return start();
  });
});


/*
Test the plugins tour feature
 */

module('tour');

asyncTest('Run tour', function() {
  var smallipopTour;
  expect(2);
  smallipopTour = $('#smallipop2');
  $('.smallipop-tour[data-smallipop-tour-index="1"]').smallipop('tour');
  return delayCall(defaultDelay, function() {
    equal($('.smallipop-tour-progress', smallipopTour).text(), '1 of 2', 'Tour should start at the selected element when no index is provided');
    $('.smallipop-tour').smallipop('tour', 2);
    return delayCall(defaultDelay, function() {
      equal($('.smallipop-tour-progress', smallipopTour).text(), '2 of 2', 'Tour with startindex 2 should start at the second element');
      return start();
    });
  });
});


watcher.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
"""
Watcher script for scss and coffeescript
By Sebastian Helzle sebastian@helzle.net
"""

import os, multiprocessing, logging

def coffee_daemon():
    logging.info('Starting: %s' % multiprocessing.current_process().name)
    
    # Start coffeescript watcher
    os.spawnlp(os.P_WAIT, 'coffee', 'coffee', '-b', '-o', 'lib/', '--watch', '-c', 'src/')
    
    logging.info('Exiting : %s' % multiprocessing.current_process().name)

def sass_daemon():
    logging.info('Starting: %s' % multiprocessing.current_process().name)
    
    # Start css watcher
    os.spawnlp(os.P_WAIT, 'sass', 'sass', '--style', 'compact', '--watch', 'scss:css')
    
    logging.info('Exiting : %s' % multiprocessing.current_process().name)

def main():
    logging.getLogger().setLevel(logging.DEBUG)
    
    coffee = multiprocessing.Process(name='coffee_daemon', target=coffee_daemon)
    coffee.daemon = True

    scss = multiprocessing.Process(name='sass_daemon', target=sass_daemon)
    scss.daemon = True
    
    try:
        coffee.start()
        scss.start()
    
        coffee.join(1)
        scss.join()
    
        logging.info('Processes ended')
    except KeyboardInterrupt:
        try:
            coffee.terminate()
            logging.info('Process coffee ended by interrupt')
        except Exception, e:
            logging.info('Processes coffee could not be terminated: %s' % e)
            
        try:
            scss.terminate()
            logging.info('Process scss ended by interrupt')
        except Exception, e:
            logging.info('Processes scss could not be terminated: %s' % e)

if __name__ == '__main__':
    main()


lib/demo.js
/*
General demo helpers for smallipop
*/

define(['jquery', 'prettify', 'piwik', 'smallipop'], function($) {
  var currentSideMenuItem, hostname, lastSideMenuY, piwikTracker, s, sideMenu, sideMenuItems, sideMenuMinY, t;
  sideMenu = $('.side-menu');
  sideMenuItems = $('a', sideMenu);
  sideMenuMinY = $('.page').position().top;
  lastSideMenuY = -1;
  currentSideMenuItem = 0;
  /* Enable prettyprint
  */

  if (typeof window.prettyPrint === "function") {
    window.prettyPrint();
  }
  /* Collapsable contentelements
  */

  $('.collapsor').click(function(e) {
    var self;
    e.preventDefault();
    self = $(this);
    return self.slideUp(300, function() {
      return $(self.attr('href')).slideDown(300);
    });
  });
  /* Smooth scroller for menu
  */

  $('.scrollToId').click(function(e) {
    e.preventDefault();
    return $('html, body').animate({
      scrollTop: $($(this).attr('href')).offset().top
    }, 800, 'swing');
  });
  /* Smallipop creation calls
  */

  $('.smallipop').smallipop();
  $('.smallipopOrange').smallipop({
    theme: 'orange'
  });
  $('.smallipopBlack').smallipop({
    theme: 'black'
  });
  $('.smallipopHideBlack').smallipop({
    theme: 'black',
    hideTrigger: true
  });
  $('.smallipopBlue').smallipop({
    theme: 'blue',
    invertAnimation: true
  });
  $('.smallipopBlueFatShadow').smallipop({
    theme: 'blue fatShadow',
    invertAnimation: true
  });
  $('.smallipopHideBlue').smallipop({
    theme: 'blue',
    hideTrigger: true,
    popupYOffset: 40,
    invertAnimation: true
  });
  $('.smallipopWhite').smallipop({
    theme: 'white'
  });
  $('.smallipopHide').smallipop({
    hideTrigger: true,
    theme: 'white',
    popupYOffset: 5
  });
  $('.smallipopHideTrans').smallipop({
    hideTrigger: true,
    theme: 'white whiteTransparent',
    popupYOffset: 20
  });
  $('.smallipopStatic').smallipop({
    theme: 'black',
    popupDistance: 0,
    popupYOffset: -14,
    popupAnimationSpeed: 100
  });
  $('.smallipopBottom').smallipop({
    theme: 'black',
    preferredPosition: 'bottom'
  });
  $('.smallipopHorizontal').smallipop({
    preferredPosition: 'right',
    theme: 'black',
    popupOffset: 10,
    invertAnimation: true
  });
  $('.smallipopFormElement').smallipop({
    preferredPosition: 'right',
    theme: 'black',
    popupOffset: 0,
    triggerOnClick: true
  });
  $('#tipcustomhint').smallipop({}, "I'm the real hint!");
  $(document).delegate('#tipkiller', 'click', function(e) {
    e.preventDefault();
    return $('#ajaxContainer').html('<div>Some new content</div>');
  });
  $(document).delegate('#tipkiller2', 'click', function(e) {
    e.preventDefault();
    return $('#ajaxContainer2').html('<div>Some new content</div>');
  });
  $('#tipChangeContent').smallipop({
    onAfterShow: function(trigger) {
      return $.smallipop.setContent(trigger, "I'm the new content and I have replaced the old boring content!");
    },
    onBeforeHide: function(trigger) {
      return $.smallipop.setContent(trigger, "Bye bye");
    }
  });
  $('#tipCSSAnimated').smallipop({
    cssAnimations: {
      enabled: true,
      show: 'animated bounceInDown',
      hide: 'animated hinge'
    }
  });
  $('#tipCSSAnimated2').smallipop({
    cssAnimations: {
      enabled: true,
      show: 'animated flipInX',
      hide: 'animated flipOutX'
    }
  });
  $('#tipCSSAnimated3').smallipop({
    cssAnimations: {
      enabled: true,
      show: 'animated fadeInLeft',
      hide: 'animated fadeOutRight'
    }
  });
  $('#tipCSSAnimated4').smallipop({
    cssAnimations: {
      enabled: true,
      show: 'animated rotateInDownLeft',
      hide: 'animated rotateOutUpRight'
    }
  });
  $('#tipDontHideOnTriggerClick').smallipop({
    hideOnTriggerClick: false
  });
  $('#tipDontHideOnContentClick').smallipop({
    hideOnPopupClick: false
  });
  /* Tour stuff
  */

  $('.smallipopTour').smallipop({
    theme: 'black wide',
    cssAnimations: {
      enabled: true,
      show: 'animated fadeIn',
      hide: 'animated fadeOut'
    }
  });
  $('#runTour').click(function() {
    return $('.smallipopTour').smallipop('tour');
  });
  $('#startExample1').click(function(e) {
    e.preventDefault();
    return $('#example2 .smallipopTour').smallipop('tour');
  });
  /* Animate smallipops when scrolling
  */

  if ($('.wobbler').length) {
    $(document).scroll(function() {
      var $win, $wobblers, scrollTop, winHeight;
      $wobblers = $('.wobbler:not(.wobble)');
      $win = $(window);
      winHeight = $win.height();
      scrollTop = $win.scrollTop();
      return $wobblers.each(function() {
        var $self, offset;
        $self = $(this);
        offset = $self.offset();
        if (offset.top > scrollTop + 50 && offset.top < scrollTop - 50 + winHeight) {
          return $self.addClass('wobble');
        }
      });
    }).trigger('scroll');
  }
  /* Special 3rd party script calls
  */

  hostname = document.location.hostname;
  if (hostname && (hostname.indexOf('github') >= 0 || hostname.indexOf('sebastianhelzle') >= 0)) {
    try {
      piwikTracker = Piwik.getTracker('https://tracking.sebastianhelzle.net/piwik.php', 5);
      piwikTracker.trackPageView();
      piwikTracker.enableLinkTracking();
    } catch (err) {

    }
    document.write(unescape("%3Cscript src='http://s7.addthis.com/js/250/addthis_widget.js#pubid=sebobo' type='text/javascript'%3E%3C/script%3E"));
    s = document.createElement('script');
    t = document.getElementsByTagName('script')[0];
    s.type = 'text/javascript';
    s.async = true;
    s.src = 'http://api.flattr.com/js/0.6/load.js?mode=auto';
    return t.parentNode.insertBefore(s, t);
  }
});


lib/main.js
var develop, hostname;

hostname = document.location.hostname;

develop = !(hostname && (hostname.indexOf('github') >= 0 || hostname.indexOf('sebastianhelzle') >= 0));

requirejs.config({
  urlArgs: develop ? 'bust=' + (new Date()).getTime() : '',
  paths: {
    jquery: "contrib/jquery.min",
    prettify: "contrib/prettify",
    smallipop: "jquery.smallipop",
    piwik: "https://tracking.sebastianhelzle.net/piwik"
  }
});

requirejs(['jquery', 'prettify', 'piwik', 'smallipop'], function($) {
  var currentSideMenuItem, err, lastSideMenuY, piwikTracker, sideMenu, sideMenuItems, sideMenuMinY;
  sideMenu = $('.side-menu');
  sideMenuItems = $('a', sideMenu);
  sideMenuMinY = $('.page').position().top;
  lastSideMenuY = -1;
  currentSideMenuItem = 0;

  /* Enable prettyprint */
  if (typeof window.prettyPrint === "function") {
    window.prettyPrint();
  }

  /* Collapsable contentelements */
  $('.collapsor').click(function(e) {
    var self;
    e.preventDefault();
    self = $(this);
    return self.slideUp(300, function() {
      return $(self.attr('href')).slideDown(300);
    });
  });

  /* Smooth scroller for menu */
  $('.scrollToId').click(function(e) {
    e.preventDefault();
    return $('html, body').animate({
      scrollTop: $($(this).attr('href')).offset().top
    }, 800, 'swing');
  });

  /* Smallipop creation calls */
  $('.smallipop').smallipop();
  $('.smallipopOrange').smallipop({
    theme: 'orange'
  });
  $('.smallipopBlack').smallipop({
    theme: 'black'
  });
  $('.smallipopHideBlack').smallipop({
    theme: 'black',
    hideTrigger: true
  });
  $('.smallipopBlue').smallipop({
    theme: 'blue',
    invertAnimation: true
  });
  $('.smallipopBlueFatShadow').smallipop({
    theme: 'blue fat-shadow',
    invertAnimation: true
  });
  $('.smallipopHideBlue').smallipop({
    theme: 'blue',
    hideTrigger: true,
    popupYOffset: 40,
    invertAnimation: true
  });
  $('.smallipopWhite').smallipop({
    theme: 'white'
  });
  $('.smallipopHide').smallipop({
    hideTrigger: true,
    theme: 'white',
    popupYOffset: 5
  });
  $('.smallipopHideTrans').smallipop({
    hideTrigger: true,
    theme: 'white white-transparent',
    popupYOffset: 20
  });
  $('.smallipopStatic').smallipop({
    theme: 'black',
    popupDistance: 0,
    popupYOffset: -14,
    popupAnimationSpeed: 100
  });
  $('.smallipopBottom').smallipop({
    theme: 'black',
    preferredPosition: 'bottom'
  });
  $('.smallipopHorizontal').smallipop({
    preferredPosition: 'right',
    theme: 'black',
    popupOffset: 10,
    invertAnimation: true
  });
  $('.smallipopFormElement').smallipop({
    preferredPosition: 'right',
    theme: 'black',
    popupOffset: 0,
    triggerOnClick: true,
    popupId: 'form'
  });
  $('#tipcustomhint').smallipop({}, "I'm the real hint!");
  $(document).delegate('#tipkiller', 'click', function(e) {
    e.preventDefault();
    return $('#ajaxContainer').html('<div>Some new content</div>');
  });
  $(document).delegate('#tipkiller2', 'click', function(e) {
    e.preventDefault();
    return $('#ajaxContainer2').html('<div>Some new content</div>');
  });
  $('#tipChangeContent').smallipop({
    onAfterShow: function(trigger) {
      return $.smallipop.setContent(trigger, "I'm the new content and I have replaced the old boring content!");
    },
    onBeforeHide: function(trigger) {
      return $.smallipop.setContent(trigger, "Bye bye");
    }
  });
  $('#tipCSSAnimated').smallipop({
    cssAnimations: {
      enabled: true,
      show: 'animated bounceInDown',
      hide: 'animated hinge'
    }
  });
  $('#tipCSSAnimated2').smallipop({
    cssAnimations: {
      enabled: true,
      show: 'animated flipInX',
      hide: 'animated flipOutX'
    }
  });
  $('#tipCSSAnimated3').smallipop({
    cssAnimations: {
      enabled: true,
      show: 'animated fadeInLeft',
      hide: 'animated fadeOutRight'
    }
  });
  $('#tipCSSAnimated4').smallipop({
    cssAnimations: {
      enabled: true,
      show: 'animated rotateInDownLeft',
      hide: 'animated rotateOutUpRight'
    }
  });
  $('#tipDontHideOnTriggerClick').smallipop({
    hideOnTriggerClick: false
  });
  $('#tipDontHideOnContentClick').smallipop({
    hideOnPopupClick: false
  });

  /* Tour stuff */
  $('.smallipopTour').smallipop({
    theme: 'black wide',
    cssAnimations: {
      enabled: true,
      show: 'animated fadeIn',
      hide: 'animated fadeOut'
    }
  });
  $('#runTour').click(function() {
    return $('.smallipopTour').smallipop('tour');
  });
  $('#startExample1').click(function(e) {
    e.preventDefault();
    return $('#example2 .smallipopTour').smallipop('tour');
  });

  /* Animate smallipops when scrolling */
  if ($('.wobbler').length) {
    $(document).scroll(function() {
      var $win, $wobblers, scrollTop, winHeight;
      $wobblers = $('.wobbler:not(.wobble)');
      $win = $(window);
      winHeight = $win.height();
      scrollTop = $win.scrollTop();
      return $wobblers.each(function() {
        var $self, offset;
        $self = $(this);
        offset = $self.offset();
        if (offset.top > scrollTop + 50 && offset.top < scrollTop - 50 + winHeight) {
          return $self.addClass('wobble');
        }
      });
    }).trigger('scroll');
  }
  if (!develop) {
    try {
      piwikTracker = Piwik.getTracker('https://tracking.sebastianhelzle.net/piwik.php', 5);
      piwikTracker.trackPageView();
      return piwikTracker.enableLinkTracking();
    } catch (_error) {
      err = _error;
    }
  }
});


lib/jquery-smallipop.js
/*!
SmallIPop 0.1.3 (11/23/2011)
Copyright (c) 2011 Small Improvements (http://www.small-improvements.com)

Licensed under the MIT (http://www.opensource.org/licenses/mit-license.php) license.

@author Sebastian Helzle (sebastian@small-improvements.com)
*/(function($) {
  $.smallipop = {
    version: '0.1.3',
    defaults: {
      popupOffset: 31,
      popupYOffset: 0,
      popupDistance: 20,
      popupDelay: 100,
      hideTrigger: false,
      theme: "default",
      infoClass: "smallipopHint",
      triggerAnimationSpeed: 150,
      popupAnimationSpeed: 200,
      invertAnimation: false,
      horizontal: false,
      preferredPosition: "top"
    },
    popup: null,
    lastId: 1,
    hideSmallipop: function() {
      var direction, shownId, sip, trigger, triggerOpt, xDistance, yDistance;
      sip = $.smallipop;
      shownId = sip.popup.data("shown");
      trigger = $(".smallipop" + shownId);
      triggerOpt = trigger.data("options") || sip.defaults;
      if (shownId && triggerOpt.hideTrigger) {
        trigger.stop(true).fadeTo(triggerOpt.triggerAnimationSpeed, 1);
      }
      direction = triggerOpt.invertAnimation ? -1 : 1;
      xDistance = sip.popup.data("xDistance") * direction;
      yDistance = sip.popup.data("yDistance") * direction;
      return sip.popup.data({
        hideDelayTimer: null,
        beingShown: false
      }).stop(true).animate({
        top: "-=" + xDistance + "px",
        left: "+=" + yDistance + "px",
        opacity: 0
      }, {
        duration: triggerOpt.popupAnimationSpeed,
        step: sip.func_ease,
        complete: function() {
          if (!$(this).data("beingShown")) {
            return $(this).css("display", "none").data("shown", "");
          }
        }
      });
    },
    killTimers: function() {
      var hideTimer, popup, showTimer;
      popup = $.smallipop.popup;
      hideTimer = popup.data("hideDelayTimer");
      showTimer = popup.data("showDelayTimer");
      if (hideTimer) {
        clearTimeout(hideTimer);
      }
      if (showTimer) {
        return clearTimeout(showTimer);
      }
    },
    func_ease: function(x, t, b, c, d) {
      if ((t /= d / 2) < 1) {
        return c / 2 * t * t + b;
      } else {
        return -c / 2 * ((--t) * (t - 2) - 1) + b;
      }
    },
    triggerMouseover: function() {
      var id, opt, popup, self, shownId, sip;
      self = $(this);
      id = self.data("id");
      sip = $.smallipop;
      popup = sip.popup;
      shownId = popup.data("shown");
      sip.killTimers();
      popup.data((id ? "triggerHovered" : "hovered"), true);
      if (id && !popup.data("beingShown") && shownId !== id) {
        opt = self.data("options");
        return popup.data("showDelayTimer", setTimeout(function() {
          var offset, popupCenter, popupH, popupOffsetLeft, popupOffsetTop, popupW, popupY, selfHeight, selfWidth, selfY, trigger, triggerOpt, win, winHeight, winWidth, windowPadding, xDistance, yDistance, yOffset;
          if (!popup.data("triggerHovered")) {
            return;
          }
          trigger = $(".smallipop" + shownId);
          triggerOpt = trigger.data("options") || sip.defaults;
          if (shownId && triggerOpt.hideTrigger) {
            trigger.stop(true).fadeTo(triggerOpt.fadeSpeed, 1);
          }
          popup.removeClass().addClass(opt.theme).data({
            beingShown: true,
            shown: id
          }).find(".sipContent").html(self.data("hint"));
          win = $(window);
          xDistance = yDistance = opt.popupDistance;
          yOffset = opt.popupYOffset;
          offset = self.offset();
          popupH = popup.outerHeight();
          popupW = popup.outerWidth();
          popupCenter = popupW / 2;
          winWidth = win.width();
          winHeight = win.height();
          selfWidth = self.outerWidth();
          selfHeight = self.outerHeight();
          windowPadding = 30;
          popupOffsetLeft = offset.left + selfWidth / 2;
          popupOffsetTop = offset.top - popupH + yOffset;
          if (opt.horizontal) {
            xDistance = 0;
            popupOffsetTop += selfHeight / 2 + popupH / 2;
            if (offset.left + selfWidth + popupW > winWidth - windowPadding) {
              popup.addClass("sipPositionedLeft");
              popupOffsetLeft = offset.left - popupW - opt.popupOffset;
              yDistance = -yDistance;
            } else {
              popup.addClass("sipPositionedRight");
              popupOffsetLeft = offset.left + selfWidth + opt.popupOffset;
            }
          } else {
            yDistance = 0;
            if (popupOffsetLeft + popupCenter > winWidth - windowPadding) {
              popupOffsetLeft -= popupCenter * 2 - opt.popupOffset;
              popup.addClass("sipAlignLeft");
            } else if (popupOffsetLeft - popupCenter < windowPadding) {
              popupOffsetLeft -= opt.popupOffset;
              popup.addClass("sipAlignRight");
            } else {
              popupOffsetLeft -= popupCenter;
            }
            selfY = offset.top - win.scrollTop();
            popupY = popupH + opt.popupDistance - yOffset;
            if ((opt.preferredPosition === "bottom" && selfY + selfHeight + popupY < winHeight - windowPadding) || selfY - popupY < windowPadding) {
              popupOffsetTop += popupH + selfHeight - 2 * yOffset;
              xDistance = -xDistance;
              yOffset = 0;
              popup.addClass("sipAlignBottom");
            }
          }
          if (opt.hideTrigger) {
            $(".smallipop" + id).stop(true).fadeTo(opt.triggerAnimationSpeed, 0);
          }
          return popup.data({
            xDistance: xDistance,
            yDistance: yDistance
          }).stop(true).css({
            top: popupOffsetTop,
            left: popupOffsetLeft,
            display: "block",
            opacity: 0
          }).animate({
            top: "-=" + xDistance + "px",
            left: "+=" + yDistance + "px",
            opacity: 1
          }, {
            duration: opt.popupAnimationSpeed,
            step: sip.func_ease,
            complete: function() {
              return popup.data("beingShown", false);
            }
          });
        }, opt.popupDelay));
      }
    },
    triggerMouseout: function() {
      var id, popup, self, sip;
      self = $(this);
      sip = $.smallipop;
      popup = sip.popup;
      id = self.data("id");
      sip.killTimers();
      popup.data((id ? "triggerHovered" : "hovered"), false);
      if (!(popup.data("hovered") || popup.data("triggerHovered"))) {
        return popup.data("hideDelayTimer", setTimeout(sip.hideSmallipop, 500));
      }
    }
  };
  return $.fn.smallipop = function(options, hint) {
    var popup, sip;
    if (options == null) {
      options = {};
    }
    if (hint == null) {
      hint = "";
    }
    sip = $.smallipop;
    options = $.extend({}, sip.defaults, options);
    if (options.moveSpeed != null) {
      options.popupAnimationSpeed = options.moveSpeed;
    }
    if (options.hideSpeed != null) {
      options.triggerAnimationSpeed = options.hideSpeed;
    }
    popup = $("#smallipop");
    if (!popup.length) {
      popup = sip.popup = $("<div id=\"smallipop\"><div class=\"sipContent\"/><div class=\"sipArrowBorder\"/><div class=\"sipArrow\"/></div>").css("opacity", 0).bind({
        mouseover: sip.triggerMouseover,
        mouseout: sip.triggerMouseout
      });
      $("body").append(popup);
      $("a", popup.get(0)).live("click", sip.hideSmallipop);
    }
    return this.each(function() {
      var newId, objHint, self;
      self = $(this);
      objHint = hint || self.attr("title") || self.find("." + options.infoClass).html();
      if (objHint && !self.hasClass("sipInitialized")) {
        newId = sip.lastId++;
        self.addClass("sipInitialized smallipop" + newId).data({
          id: newId,
          options: options,
          hint: objHint
        }).attr("title", "").bind({
          mouseover: sip.triggerMouseover,
          mouseout: sip.triggerMouseout,
          click: sip.hideSmallipop
        });
        return $("a", this).live("click", sip.hideSmallipop);
      }
    });
  };
})(jQuery);